packages/ics721/src/execute.rs

```diff
@@ -410,12 +414,17 @@ where
             return Err(ContractError::NotEscrowedByIcs721(access.owner));
         }
 
-        // cw721 doesn't support on-chain metadata yet
-        // here NFT is transferred to another chain, NFT itself may have been transferred to his chain before
+        // here NFT was transferred before, in this case it is stored in the storage, otherwise this is the home chain,
+        // and the NFT is transferred for the first time and onchain data comes from the cw721 contract
         // in this case ICS721 may have metadata stored
-        let token_metadata = TOKEN_METADATA
+        let token_metadata = match IBC_RECEIVE_TOKEN_METADATA
             .may_load(deps.storage, (class.id.clone(), token_id.clone()))?
-            .flatten();
+            .flatten()
+        {
+            Some(metadata) => Some(metadata),
+            // incase there is none in the storage, this is the 'home' chain, so metadata is retrieved from the cw721 contract
+            None => info.extension.map(|ext| to_json_binary(&ext)).transpose()?,
```

Comment:
Here, on source chain, nft extension (metadata) is retrieved and passed in ibc packet.